client/starwhale/core/job/scheduler.py

```diff
@@ -0,0 +1,258 @@
+import os
+import time
+import typing as t
+import threading
+import concurrent.futures
+from abc import abstractmethod
+from pathlib import Path
+from multiprocessing.connection import Connection
+
+from loguru import logger
+
+from starwhale.consts import EvaluationResultKind
+from starwhale.api._impl import wrapper
+from starwhale.utils.config import SWCliConfigMixed
+from starwhale.core.job.model import Step, Task, STATUS
+from starwhale.api._impl.wrapper import (
+    EvaluationQuery,
+    EvaluationMetric,
+    EvaluationResult,
+)
+
+
+class TaskPipe:
+    def __init__(self, idx: str, main_conn: Connection, sub_conn: Connection):
+        self.id = idx
+        self.main_conn = main_conn
+        self.sub_conn = sub_conn
+
+
+def _init_datastore(root_path: str, project: str, eval_id: str) -> wrapper.Evaluation:
+    os.environ["SW_ROOT_PATH"] = root_path
+    os.environ["SW_PROJECT"] = project
+    os.environ["SW_EVAL_ID"] = eval_id
+    logger.debug("datastore info:{}, {}, {}", root_path, project, eval_id)
+    return wrapper.Evaluation()
+
+
+class Scheduler:
+    def __init__(
+        self,
+        project: str,
+        version: str,
+        module: str,
+        workdir: Path,
+        src_dir: Path,
+        dataset_uris: t.List[str],
+        steps: t.List[Step],
+        **kw: t.Any,
+    ):
+        self.project = project
+        self.steps = steps
+        self.dataset_uris = dataset_uris
+        self.module = module
+        self.workdir = workdir
+        self.src_dir = src_dir
+        self.version = version
+        self.__split_tasks(**kw)
+        self._lock = threading.Lock()
+        self._sw_config = SWCliConfigMixed()
+        self._datastore = _init_datastore(
+            str(self._sw_config.datastore_dir), project, version
+        )
+
+    def add_data(self, data: t.Any) -> None:
+        if isinstance(data, EvaluationResult):
+            self._datastore.log_result(
+                data_id=data.data_id, result=data.result, **data.kwargs
+            )
+        elif isinstance(data, EvaluationMetric):
+            self._datastore.log_metrics(metrics=data.metrics, **data.kwargs)
+
+    def query_data(self, query: EvaluationQuery) -> t.Any:
+        logger.debug("main receive query data:{}", query.kind)
+        if query.kind == EvaluationResultKind.RESULT:
+            logger.debug(
+                "hi,all result size: {}",
+                len([res for res in self._datastore.get_results()]),
```

Comment:
`len([res for res in self._datastore.get_results()])` -> `len(list(self._datastore.get_results()))` ?